packages/sitecore-jss-nextjs/src/services/graphql-sitemap-service.ts

```diff
@@ -1,258 +1,259 @@
+  ): Promise<StaticPath[]> {
+    if (!language) {
+      throw new RangeError('The language value must be a non-empty string');
+    }
 
-          return null;
-        });
+    const results: StaticPath[] = [];
+    let hasNext = true;
+    let after = '';
 
-      const items = data?.search.results ? data.search.results : [];
+    while (hasNext) {
+      const fetchResponse = await client.request<SitePageQueryResult>(query, {
+        rootItemId,
+        language,
+        pageSize: this.options.pageSize,
+        after,
+      });
 
-      // todo: convert to array.map (Anastasiya, March 2021)
-      const staticPaths = items.reduce((list: StaticPath[], item: { url: { path: string } }) => {
+      // transform to array of paths
+      fetchResponse?.search?.results.forEach((item: { url: { path: string } }) => {
         // replace first and last /
-        // transform to array of paths
         const path = item.url.path.replace(/^\/|\/$/g, '').split('/');
-
-        return list.concat(formatStaticPath(path, locale));
+        results.push(formatStaticPath(path, language));
       }, []);
 
-      return staticPaths;
-    };
-
-    // Fetch paths using all locales
-    const paths = await Promise.all(locales.map(getStaticPaths));
-
-    // merge promises result
-    return ([] as StaticPath[]).concat(...paths);
-  }
-
-  /**
-   * Returns new graphql client instance
-   */
-  private createClient(): GraphQLRequestClient {
-    const { endpoint } = this.config;
+      hasNext = fetchResponse.search.pageInfo.hasNext;
+      after = fetchResponse.search.pageInfo.endCursor;
+    }
 
-    return new GraphQLRequestClient(endpoint);
+    return results;
   }
 
   /**
    * Returns query to request root item id
    * @param {string} rootItemPath root item path
    */
   private getRootItemIdQuery(rootItemPath: string): string {
-    return `query {
+    // TODO: this should use 'layout' query, which only needs the site name (Anastasiya, March 2021)
+    return `query RootItemQuery{
       item(path:"${rootItemPath}") {
         id
       }
     }`;
   }
 
   /**
-   * Request root item id using path
+   * Request root item id using the item path
+   * @param {GraphQLRequestClient} client that fetches data from a GraphQL endpoint.
    * @param {string} rootItemPath root item path
    */
-  private async getRootItemId(rootItemPath: string): Promise<string | undefined> {
+  private async getRootItemId(client: GraphQLRequestClient, rootItemPath: string): Promise<string> {
     const query = this.getRootItemIdQuery(rootItemPath);
-
-    const data = await this.createClient()
-      .request<{ item: { id: string } }>(query)
-      .catch((error) => {
-        console.error(
-          chalk.red('Error occurred while fetching root item id:'),
-          error.response || error
-        );
-
-        return null;
-      });
-
+    const data = await client.request<{ item: { id: string } }>(query);
```

Comment:
@anastasiya29 You have type `{ item: { id: string } }`, but you check `data?.item?.id`.
In this case `item` is optional? Can you align it?

Reply:
Good catch. Done.